component/libfuse/libfuse_wrapper.h

```diff
@@ -286,4 +294,187 @@ static int fill_dir_entry(fuse_fill_dir_t filler, void *buf, char *name, stat_t
+    if (fobj) {
+        memset(fobj, 0, sizeof(file_handle_t));
+        fobj->fd = fd;
+        fobj->obj = obj;
+    }
+
+    return fobj;
+}
+
+// release_native_file_object : Release the native C-struct for handle 
+static void release_native_file_object(fuse_file_info_t* fi)
+{
+    // Called on close operation from libfuse component
+    file_handle_t* handle_obj = (file_handle_t*)fi->fh;
+    if (handle_obj) {
+        free(handle_obj);
+    }
+}
+
+
+// native_pread :  Do pread on file directly without involving any Go code
+static int native_pread(char *path, char *buf, size_t size, off_t offset, file_handle_t* handle_obj)
+{
+    errno = 0;
+    int res = pread(handle_obj->fd, buf, size, offset);
+    if (res == -1)
+        res = -errno;
+        
+    return res;
+}
+
+// native_pwrite :  Do pwrite on file directly without involving any Go code
+static int native_pwrite(char *path, char *buf, size_t size, off_t offset, file_handle_t* handle_obj)
+{
+    errno = 0;
+    int res = pwrite(handle_obj->fd, buf, size, offset);
+    if (res == -1)
+        res = -errno;
+
+    // Increment the operation counter and mark a write was done on this handle
+    handle_obj->dirty = 1;
+    handle_obj->cnt++;
+    if (!(handle_obj->cnt % CACHE_UPDATE_COUNTER)) {
+        // Time to send a call up to update the cache
+        blobfuse_cache_update(path);
+        handle_obj->cnt = 0;
+    }
+
+    return res;
+}
+
+// native_read_file : Read callback to decide whether to natively read or punt call to Go code
+static int native_read_file(char *path, char *buf, size_t size, off_t offset, fuse_file_info_t *fi)
+{
+    file_handle_t* handle_obj = (file_handle_t*)fi->fh;
+    
+    if (handle_obj->fd == 0) {
+        return libfuse_read(path, buf, size, offset, fi);
+    }
+
+    return native_pread(path, buf, size, offset, handle_obj);
+}
+
+// native_write_file : Write callback to decide whether to natively write or punt call to Go code
+static int native_write_file(char *path, char *buf, size_t size, off_t offset, fuse_file_info_t *fi)
+{
+    file_handle_t* handle_obj = (file_handle_t*)fi->fh;
+    
+    if (handle_obj->fd == 0) {
+        return libfuse_write(path, buf, size, offset, fi);
+    }
+    
+    return native_pwrite(path, buf, size, offset, handle_obj);
+}
+
+// native_flush_file : Flush the file natively and call flush up in the pipeline to upload this file
+static int native_flush_file(char *path, fuse_file_info_t *fi)
+{
+    file_handle_t* handle_obj = (file_handle_t*)fi->fh;
+    int ret = libfuse_flush(path, fi);
+    if (ret == 0) {
+        // As file is flushed and uploaded, reset the dirty bit here
+        handle_obj->dirty = 0;
+    }
+
+    return ret;
+}
+
+
+#ifdef ENABLE_READ_AHEAD
+// read_ahead_handler : Method to serve read call from read-ahead buffer if possible
+static int read_ahead_handler(char *path, char *buf, size_t size, off_t offset, file_handle_t* handle_obj) 
```

Comment:
so we're not shipping this, correct?

Reply:
Yes read-ahead code is removed. However, I left this function in code to just preserve the logic used here for read-ahead. This is disabled as of now and the function is not called from anywhere. Just preserved the logic that's all, if suggested, I can remove this piece as well.